### PR TITLE
[WIP] Hyper CLI

### LINKED ISSUE #2288: Hyper CLI
As started in https://github.com/zeit/hyper/pull/282, we will in 2.0 have a way to manage hyper via the CLI.

### PR DESCRIPTION
**Pull request by chabou**
Will fix #2288

### CODE REVIEW

**Comment by chabou** [comment]
* Hyper CLI javascript file is bootstrapped by scripts (located in `build/(mac|win|linux)/`)
* Right platform script(s) are copied at build time to `bin/`
* At Hyper startup time (only in production mode), platform script is symlinked to `/usr/local/bin` (or `/usr/bin/` on Linux)

For now, only macOS has been tested and it works great:
* Build Hyper `yarn dist -- -m` 
* Launch `./dist/mac/Hyper.app`
* Type `hyper <args>`
* <args> are printed

File: app/utils/cli-install.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+const lstat = pify(fs.lstat);
+const readlink = pify(fs.readlink);
+const unlink = pify(fs.unlink);
+const symlink = pify(fs.symlink);
+
+const target = process.platform === 'darwin' ? '/usr/local/bin/hyper' : '/usr/bin/hyper';
+const source = cliScriptPath;
+
+const checkInstall = () => {
```
Comment (albinekb) [inline app/utils/cli-install.js]: I would write all these promise-chain-functions without the return statement (implicit return) saves a few lines and also makes the code more readable, but if it's like this in the readme, keep it. 👍  (very very minor thing 😂 )

File: app/index.js [hunk 2]
```diff
@@ -97,6 +98,13 @@ if (isDev) {
 } else {
   //eslint-disable-next-line no-console
   console.log('running in prod mode');
+  if (process.platform === 'win32') {
+    //eslint-disable-next-line no-console
+    addBinToUserPath().catch(err => console.error('Failed to add Hyper CLI path to user PATH', err));
```
Comment (albinekb) [inline app/index.js]: Should this be shown to the user (notification) or not?

File: app/index.js [hunk 3]
```diff
@@ -97,6 +98,13 @@ if (isDev) {
...
+  } else {
+    //eslint-disable-next-line no-console
+    addSymlink().catch(err => console.error('Failed to symlink Hyper CLI', err));
```
Comment (albinekb) [inline app/index.js]: Same with this one, shown to the user or just a log?

File: app/package.json [hunk 4]
```diff
@@ -27,6 +27,7 @@
     "queue": "4.4.0",
     "semver": "5.4.1",
     "shell-env": "0.3.0",
+    "util.promisify": "false1.0.0",
```
Comment (albinekb) [inline app/package.json]: this is built in: https://nodejs.org/api/util.html#util_util_promisify_original

File: app/utils/cli-install.js (hunk 1 above)
Comment (chabou) [reply app/utils/cli-install.js]: I'll change that

File: app/index.js (hunk 2 above)
Comment (chabou) [reply app/index.js]: Not enough place to print error in notification.
On macOS, error could be read in Console.app. But I don't know for other platform

File: app/package.json (hunk 4 above)
Comment (chabou) [reply app/package.json]: I forgot to remove it from this file.
I replaced `util.promisify` by `pify`.
Promisify built-in is only available from node v8.x and we are using v7.9.0 (with Electron v1.7.8).

File: app/index.js (hunk 2 above)
Comment (albinekb) [reply app/index.js]: Let's keep it as it is and see if it's needed later on 👍  Let's not over engineer it

File: app/package.json (hunk 4 above)
Comment (albinekb) [reply app/package.json]: 👍  No probs

File: app/utils/cli-install.js (hunk 1 above)
Comment (chabou) [reply app/utils/cli-install.js]: I really prefer a first explicit return in these functions

**Comment by chabou** [comment]
I'm closing this PR.
As @albinekb asked, I'm creating a feature branch with multiple PR to make review easier.

**Pull request closed without merging**